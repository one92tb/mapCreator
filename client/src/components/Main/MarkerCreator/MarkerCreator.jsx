import React, { Component } from "react";
import { postRecord } from "../../../actions/postRecord";
import { getSelectedMarker } from "../../../actions/getSelectedMarker";
import { removeRecord } from "../../../actions/removeRecord";
import { editRecord } from "../../../actions/editRecord";
import { connect } from "react-redux";
import {
  Wrapper,
  Inner,
  ImageInsideMarker,
  ImageWithoutMarker,
  Form,
  FormGroup,
  LabelColor,
  LabelFile,
  LabelName,
  HideInput,
  Input,
  InputFile,
  SubmitBtn,
  RemoveBtn,
  DownloadBtn,
  MarkerIcon,
  UploadButton,
  CustomButton,
  ImageBox,
  AdditionalWrapper,
  MarkerIconBox,
  ButtonGroup,
  ErrorMessage
} from "./style";


const domtoimage = require("dom-to-image");

class MarkerCreator extends Component {
  constructor(props) {
    super(props);
    this.myRef = React.createRef();
    this.textInput = React.createRef();
    this.state = {
      markerName: "",
      markerImageFile: "",
      displaySelectedImage: "IMG-default.png",
      color: "#000",
      uploadStatus: false,
      markerNameError: "",
      markerImageFileError: ""
    };
  }

  componentDidUpdate(prevProps) {
    const { selectedMarker } = this.props;
    if (selectedMarker.id !== prevProps.selectedMarker.id) {
      this.setState({
        markerName: selectedMarker.name,
        displaySelectedImage: selectedMarker.url,
        markerImageFile: ""
      });
    }
  }

  onChange = event => {
    const { markerImageFile } = this.state;

    if (event.target.name === "markerName") {
      this.setState({ markerName: event.target.value });
      if (event.target.value.length > 2) {
        this.setState({
          markerNameError: ""
        });
      }
    } else if (event.target.name === "color") {
      this.setState({ color: event.target.value });
    } else {
      if (event.target.files[0]) {
        this.setState(
          {
            markerImageFile: event.target.files[0],
            displaySelectedImage: URL.createObjectURL(event.target.files[0])
          },
          () =>
            /\.(png)$/i.test(markerImageFile.name) &&
            this.setState({ markerImageFileError: "" })
        );
      }
    }
  };

  validate = () => {
    const { markerName, markerImageFile } = this.state;
    const { selectedMarker } = this.props;

    let isError = false;
    const errors = {
      markerNameError: "",
      markerImageFileError: ""
    };

    const validateArray = [
      {
        condition: markerName.length < 3,
        nameOfErrorProperty: "markerNameError",
        messageError: "Username needs to be atleast 3 characters long"
      },
      {
        condition: !/\.(png)$/i.test(markerImageFile.name) && markerImageFile,
        nameOfErrorProperty: "markerImageFileError",
        messageError: "format must be .png"
      },
      {
        condition: !markerImageFile && !selectedMarker.id,
        nameOfErrorProperty: "markerImageFileError",
        messageError: "Input file cannot be empty"
      }
    ];

    validateArray.forEach(validate => {
      if (validate.condition) {
        isError = true;
        errors[validate.nameOfErrorProperty] = validate.messageError;
      }
    });

    this.setState({
      ...this.state,
      ...errors
    });

    return isError;
  };

  sendRecord = event => {
    event.preventDefault();
    const { postRecord, editRecord, selectedMarker } = this.props;
    const { markerImageFile, markerName } = this.state;
    const err = this.validate();

    let fd = new FormData();

    fd.append("file", markerImageFile);
    fd.append("markerName", markerName);

    if (!err) {
      if (selectedMarker.id && !selectedMarker.isDeleted) {
        if (markerImageFile === "") {
          editRecord(
            {
              name: markerName,
              icon: selectedMarker.icon
            },
            selectedMarker.id
          );
        } else {
          console.log("z pliku");
          editRecord(fd, selectedMarker.id);
        }
      } else {
        postRecord(fd);

        this.setState({
          markerName: "",
          markerImageFile: "",
          displaySelectedImage: "IMG-default.png"
        });
      }
    }
  };

  downloadMarker = () => {
    const { markerName } = this.state;
    const err = this.validate();

    if (!err) {
      const node = this.myRef;
      domtoimage.toPng(node).then(dataUrl => {
        var link = document.createElement("a");
        link.download = `${markerName}.png`;
        link.href = dataUrl;
        link.click();
      });
    }
  };

  removeRecord = () => {
    const { removeRecord, getSelectedMarker, selectedMarker } = this.props;

    this.setState({
      markerName: "",
      markerImageFile: "",
      displaySelectedImage: "IMG-default.png"
    });

    removeRecord(selectedMarker.id);

    getSelectedMarker({
      ...selectedMarker,
      isDeleted: true
    });
  };

  handleUpload = status => {
    this.setState({
      uploadStatus: status
    });
  };

  render() {
    const {
      uploadStatus,
      displaySelectedImage,
      color,
      markerName,
      markerImageFile,
      markerNameError,
      markerImageFileError
    } = this.state;
    const { selectedMarker } = this.props;

    return (
      <Wrapper>
        <Inner>
          <ButtonGroup>
            <UploadButton
              status={uploadStatus}
              onClick={() => this.handleUpload(true)}
            >
              Upload Marker
            </UploadButton>
            <CustomButton
              status={uploadStatus}
              onClick={() => this.handleUpload(false)}
            >
              Custom Marker
            </CustomButton>
          </ButtonGroup>
          {uploadStatus ? (
            <ImageBox>
              <ImageWithoutMarker alt="" src={displaySelectedImage} />
            </ImageBox>
          ) : (
            <AdditionalWrapper>
              <MarkerIconBox
                innerRef={div => {
                  this.myRef = div;
                }}
              >
                <MarkerIcon background={color}>
                  <ImageInsideMarker src={displaySelectedImage} />
                </MarkerIcon>
              </MarkerIconBox>
            </AdditionalWrapper>
          )}

          <Form onSubmit={this.sendRecord}>
            <FormGroup>
              <LabelName htmlFor="exampleText">Name</LabelName>
              <Input
                type="text"
                name="markerName"
                value={markerName}
                onChange={e => this.onChange(e)}
              />
              {markerNameError && (
                <ErrorMessage>{markerNameError}</ErrorMessage>
              )}
            </FormGroup>
            {!uploadStatus && (
              <FormGroup>
                <LabelColor htmlFor="exampleColor">Color</LabelColor>
                <Input
                  onChange={e => this.onChange(e)}
                  type="color"
                  name="color"
                  id="exampleColor"
                  value={color}
                  placeholder="color placeholder"
                />
              </FormGroup>
            )}
            <FormGroup>
              <LabelFile htmlFor="file">
                <InputFile>Choose file to send</InputFile>{" "}
                {displaySelectedImage === "IMG-default.png"
                  ? "Not file detected"
                  : markerImageFile === ""
                    ? `${markerName}.png`
                    : markerImageFile.name}
              </LabelFile>
              <HideInput
                type="file"
                id="file"
                name="markerImage"
                onChange={e => this.onChange(e)}
              />
              {markerImageFileError && (
                <ErrorMessage>{markerImageFileError}</ErrorMessage>
              )}
            </FormGroup>
            {uploadStatus ? (
              <FormGroup>
                <SubmitBtn>
                  {selectedMarker.id && !selectedMarker.isDeleted
                    ? "Edit Marker"
                    : "Upload new marker"}
                </SubmitBtn>
                {selectedMarker.id && !selectedMarker.isDeleted ? (
                  <RemoveBtn onClick={this.removeRecord} type="button">
                    Remove Marker
                  </RemoveBtn>
                ) : (
                  ""
                )}
              </FormGroup>
            ) : (
              <FormGroup>
                <DownloadBtn onClick={this.downloadMarker} type="button">
                  Download
                </DownloadBtn>
              </FormGroup>
            )}
          </Form>
        </Inner>
      </Wrapper>
    );
  }
}

const mapDispatchToProps = {
  postRecord,
  removeRecord,
  editRecord,
  getSelectedMarker
};

const mapStateToProps = state => ({
  selectedMarker: state.selectedMarker.selectedMarker
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MarkerCreator);
