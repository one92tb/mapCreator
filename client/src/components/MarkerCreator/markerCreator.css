.markerImage {
  width: 50px;
  height: 50px;
  display: block;
  text-align: center;
}

.imageBox {
  border: 1px solid #000;
  width: 100px;
  height: 100px;
  margin-left: auto;
  margin-right: auto;
  display: flex;
  align-items: center;
  justify-content: center;
}

.wrapper {
  margin-top: 30px;
}

.markerForm {
  display: block;
}
.addBtn {
  margin-right: 10px;
}
.markerInputFile {
  visibility: hidden;
  position: absolute;
  opacity: 0;
  z-index: 99;
}
.inputLabel{
  background: blue;
  padding: 10px;
  color: #fff;

}
