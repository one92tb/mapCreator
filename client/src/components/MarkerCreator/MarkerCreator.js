import React, {Component} from 'react';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
  FormText
} from 'reactstrap';
import './markerCreator.css'

class MarkerCreator extends Component {
  constructor(props){
    super(props);
    this.state = {
      selectedFile: null
    }

    this.fileSelectedHandler = this.fileSelectedHandler.bind(this);
  }

  fileSelectedHandler(e){
    console.log(e.target.files[0]);
  }

  render() {
    console.log(this.state);
    return (<div>
      <div className="markerImage">

      </div>
      <Form>
        <FormGroup>
          <Label for="exampleText">Name</Label>
          <Input type="text" name="text" id="exampleText"/>
        </FormGroup>
        <FormGroup>
          <Input type="file" name="file" id="exampleFile" onChange={(e) => this.fileSelectedHandler(e)}/>
          <FormText color="muted"></FormText>
        </FormGroup>
        <Button>Upload</Button>
      </Form>
    </div>);
  }
}

export default MarkerCreator;
