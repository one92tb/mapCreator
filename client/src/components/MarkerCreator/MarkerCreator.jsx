import React, {Component} from 'react';
import {postRecord} from '../../actions/postRecord';
import {connect} from 'react-redux';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
  FormText
} from 'reactstrap';
import './markerCreator.css'

class MarkerCreator extends Component {
  constructor(props) {
    super(props);
    this.state = {
      markerName: '',
      markerImage: ''
    }
  }

  onChange = (event) => {
    if (event.target.name === "markerName") {
      this.setState({markerName: event.target.value})
    } else {
      this.setState({markerImage: event.target.files[0]})
    }
  }

  sendRecord = (event) => {
    event.preventDefault();
    const postRecord = this.props.postRecord;

    const fd = new FormData();

    fd.append('file', this.state.markerImage);
    fd.append('markerName', this.state.markerName);

    postRecord(fd);
  }

  render() {
    console.log(this.state);
    return (<div>
      <div className="markerImage"></div>
      <Form onSubmit={this.sendRecord}>
        <FormGroup>
          <Label for="exampleText">Name</Label>
          <Input type="text" name="markerName" onChange={(e) => this.onChange(e)}/>
        </FormGroup>
        <FormGroup>
          <Input type="file" name="markerImage" onChange={(e) => this.onChange(e)}/>
          <FormText color="muted"></FormText>
        </FormGroup>
        <Button>Upload</Button>
      </Form>
    </div>);
  }
}



const mapDispatchToProps = {
  postRecord
}

export default connect(null, mapDispatchToProps)(MarkerCreator);
