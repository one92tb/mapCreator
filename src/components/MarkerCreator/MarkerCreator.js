import React, {Component} from 'react';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
  FormText
} from 'reactstrap';
import './markerCreator.css'

class MarkerCreator extends Component {
  render() {
    return (<div>
      <div className="markerImage">
      
      </div>
      <Form>
        <FormGroup>
          <Label for="exampleText">Name</Label>
          <Input type="text" name="text" id="exampleText"/>
        </FormGroup>
        <FormGroup>
          <Input type="file" name="file" id="exampleFile"/>
          <FormText color="muted"></FormText>
        </FormGroup>
      </Form>
    </div>);
  }
}

export default MarkerCreator;
