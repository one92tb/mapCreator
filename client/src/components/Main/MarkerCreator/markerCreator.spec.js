import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { MarkerCreator } from "./MarkerCreator";

let domtoimage = require("dom-to-image");

describe("marker creator component", () => {
  let wrapper;

  it("passess all props to markerCreator component", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("componentDidMount", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      id: 1,
      name: "test1",
      icon: "1548784752159.png",
      userId: 1,
      url: "http://localhost:8080/images/1552937459493.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    expect(fetchMarkers).toHaveBeenCalledTimes(1);
  });

  it("componentDidUpdate", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    const SelectedNewMarker = {
      icon: "1550173679016.png",
      id: 2,
      name: "basen",
      userId: 1,
      url: "http://localhost:8080/images/1550173679016.png"
    };
    wrapper.setProps({
      selectedMarker: SelectedNewMarker
    });

    expect(wrapper.state().markerName).toEqual(SelectedNewMarker.name);
    expect(wrapper.state().displaySelectedImage).toEqual(SelectedNewMarker.url);
    expect(wrapper.state().markerImageFile).toEqual("");
  });

  it("on change", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    const markerInput = wrapper.find("input").first();
    const fileInput = wrapper.find("input").at(1);
    const file = {
      name: "someFile.png"
    };

    global.URL.createObjectURL = jest.fn(() => "details");

    wrapper.setState({
      markerNameError: "someError",
      markerImageFileError: "someError2"
    });

    markerInput.simulate("change", {
      target: {
        name: "markerName",
        value: "test1"
      }
    });

    fileInput.simulate("change", {
      target: {
        name: "markerImage",
        files: [file]
      }
    });

    expect(file.name).toMatch(/\.(png)$/i);
    expect(wrapper.state().markerNameError).toEqual("");
    expect(wrapper.state().markerImageFileError).toEqual("");
    expect(wrapper.state().markerName).toEqual("test1");
    expect(wrapper.state().markerImageFile.name).toEqual("someFile.png");
  });

  it("marker success sending", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = "";
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "test11",
      markerImageFile: {
        name: "someFile.png"
      }
    });

    const form = wrapper.find("form");

    form.simulate("submit", { preventDefault() {} });

    expect(wrapper.state().markerNameError).toBe("");
    expect(wrapper.state().markerImageFileError).toBe("");
  });

  it("marker failure sending - wrong file format", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = "";
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "test11",
      markerImageFile: {
        name: "someFile"
      }
    });

    const form = wrapper.find("form");

    form.simulate("submit", { preventDefault() {} });

    expect(wrapper.state().markerImageFileError).toBe("Format must be .png");
  });

  it("marker failure sending - without image file and marker name", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = "";
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({});

    const form = wrapper.find("form");

    form.simulate("submit", { preventDefault() {} });

    expect(wrapper.state().markerNameError).toBe(
      "Username needs to be atleast 3 characters long"
    );
    expect(wrapper.state().markerImageFileError).toBe(
      "Input file cannot be empty"
    );
  });

  it("marker failure sending - too short marker name", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = "";
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "t1",
      markerImageFile: {
        name: "someFile.png"
      }
    });

    const form = wrapper.find("form");

    form.simulate("submit", { preventDefault() {} });

    expect(wrapper.state().markerNameError).toBe(
      "Username needs to be atleast 3 characters long"
    );
  });

  it("marker failure sending - this marker name is exist", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = "";
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "test1",
      markerImageFile: {
        name: "someFile"
      }
    });

    const form = wrapper.find("form");

    form.simulate("submit", { preventDefault() {} });
    expect(wrapper.state().markerNameError).toBe(
      "This marker is already exist"
    );
  });

  it("edit marker success - change only marker name", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };

    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "editedName"
    });

    const form = wrapper.find("form");
    form.simulate("submit", { preventDefault() {} });

    expect(wrapper.state().markerNameError).toBe("");
    expect(wrapper.state().markerImageFileError).toBe("");
  });

  it("edit marker success - change marker name and image", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };

    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "editedName",
      markerImageFile: {
        name: "someFile.png"
      }
    });

    const form = wrapper.find("form");
    form.simulate("submit", { preventDefault() {} });

    expect(wrapper.state().markerNameError).toBe("");
    expect(wrapper.state().markerImageFileError).toBe("");
  });

  it("remove marker", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      markerName: "test1",
      markerImageFile: {
        name: "1548784752159.png"
      }
    });

    const removeBtn = wrapper.find("button").at(3);
    removeBtn.simulate("click");

    expect(wrapper.instance().state.markerName).toBe("");
    expect(wrapper.instance().state.markerImageFile).toBe("");
    expect(wrapper.instance().state.displaySelectedImage).toBe(
      "IMG-default.png"
    );
    expect(removeMarker).toHaveBeenCalledTimes(1);
    expect(getSelectedMarker).toHaveBeenCalledTimes(1);
  });
  it("change upload interface to custom", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    const customBtn = wrapper.find("button").at(1);
    customBtn.simulate("click");
    expect(wrapper.instance().state.uploadStatus).toBe(false);
  });

  it("change custom interface to upload", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      uploadStatus: false
    });
    const customBtn = wrapper.find("button").at(0);
    customBtn.simulate("click");

    expect(wrapper.instance().state.uploadStatus).toBe(true);
  });

  it("download image - success", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      uploadStatus: false,
      markerName: "test1"
    });

    const DownloadBtn = wrapper.find("button").at(2);

    domtoimage.toPng = jest.fn(arg => {
      let promise = new Promise((resolve, reject) => {
        resolve("myblob");
      });
      return promise;
    });

    DownloadBtn.simulate("click");

    expect(domtoimage.toPng).toHaveBeenCalledTimes(1);
    expect(wrapper.instance().state.markerNameError).toBe("");
    expect(wrapper.instance().state.markerImageFile).toBe("");
  });

  it("download image - failure", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = shallow(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );

    wrapper.setState({
      uploadStatus: false,
      markerName: "t1"
    });

    const DownloadBtn = wrapper.find("button").at(2);

    const mockFn = jest.fn(arg => {
      let promise = new Promise((resolve, reject) => {
        resolve("myblob");
      });
      return promise;
    });

    domtoimage.toPng = mockFn;

    DownloadBtn.simulate("click");

    expect(mockFn).toHaveBeenCalledTimes(0);
    expect(wrapper.instance().state.markerNameError).toBe(
      "Username needs to be atleast 3 characters long"
    );
    expect(wrapper.instance().state.markerImageFile).toBe("");
  });

  it("imageBox - ref", () => {
    const postMarker = jest.fn();
    const removeMarker = jest.fn();
    const editMarker = jest.fn();
    const getSelectedMarker = jest.fn();
    const fetchMarkers = jest.fn();
    const selectedMarker = {
      icon: "1548784752159.png",
      id: 1,
      name: "test1",
      userId: 1,
      url: "http://localhost:8080/images/1548784752159.png"
    };
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    wrapper = mount(
      <MarkerCreator
        markers={markers}
        selectedMarker={selectedMarker}
        getSelectedMarker={getSelectedMarker}
        editMarker={editMarker}
        postMarker={postMarker}
        removeMarker={removeMarker}
        fetchMarkers={fetchMarkers}
      />
    );
    wrapper.setState({
      uploadStatus: false
    });

    expect(wrapper.instance().imageBox).toBeTruthy();
    expect(wrapper.instance().inputFile).toBeTruthy();
  });
});
