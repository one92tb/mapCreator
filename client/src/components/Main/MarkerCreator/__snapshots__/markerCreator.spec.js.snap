// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`marker creator component passess all props to markerCreator component 1`] = `
<div>
  <div>
    <div>
      <button
        onClick={[Function]}
        status={true}
      >
        Upload Marker
      </button>
      <button
        onClick={[Function]}
        status={true}
      >
        Custom Marker
      </button>
    </div>
    <div>
      <img
        alt=""
        src="IMG-default.png"
      />
    </div>
    <form
      onSubmit={[Function]}
    >
      <div>
        <label
          htmlFor="exampleText"
        >
          Name
        </label>
        <input
          name="markerName"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          htmlFor="file"
        >
          <span>
            Choose file to send
          </span>
           
          Not file detected
        </label>
        <input
          id="file"
          name="markerImage"
          onChange={[Function]}
          type="file"
          value=""
        />
      </div>
      <div>
        <button>
          Edit Marker
        </button>
        <button
          onClick={[Function]}
          type="button"
        >
          Remove Marker
        </button>
      </div>
    </form>
  </div>
</div>
`;
